kolibri/core/assets/src/api_resource.js

```diff
@@ -0,0 +1,263 @@
+const logging = require('loglevel');
+const rest = require('rest');
+const mime = require('rest/interceptor/mime');
+
+const client = rest.wrap(mime);
+
+/** Class representing a single API resource object */
+class Model {
+  /**
+   * Create a model instance.
+   * @param {object} data - data to insert into the model at creation time - should include at
+   * least an id for fetching.
+   * @param {Resource} resource - object of the Resource class, specifies the urls and fetching
+   * behaviour for the model.
+   */
+  constructor(data, resource) {
+    // Assign any data to the attributes property of the Model.
+    this.attributes = {};
+    Object.assign(this.attributes, data);
+    this.resource = resource;
+    if (!this.resource) {
+      throw new TypeError('resource must be defined');
+    }
+    // Register any created model with the Resource to track model instances.
+    this.resource.addModel(this);
+    this.synced = false;
+  }
+
+  /**
+   * Method to fetch data from the server for this particular model.
+   * Takes no parameters, as all specification should be in the associated Resource object.
+   * @returns {Promise} - Promise is resolved with Model attributes when the XHR successfully
+   * returns, otherwise reject is called with the response object.
+   */
+  fetch() {
+    return new Promise((resolve, reject) => {
+      // Do a fetch on the URL.
+      client({ path: this.url }).then((response) => {
+        // Set the retrieved Object onto the Model instance.
+        this.set(response.entity);
+        // Flag that the Model has been fetched.
+        this.synced = true;
```

Comment:
This seems possibly brittle, given all the async stuff going on? How exactly will it be getting used? And, for e.g., what happens when it's fetched a second time?